File: backend/app/api/v1/endpoints/faq_categories.py
"""
FAQ Categories API endpoints
Following the FAQ endpoints pattern from CLAUDE.md specifications.
"""
# Standard library imports
from typing import List, Optional

# Third-party imports
from fastapi import APIRouter, Depends, Query, HTTPException

# Local imports
from app.core.deps import get_admin_user, get_current_user
from app.models.user import User
from app.schemas.faq_category import (
    FAQCategoryCreate,
    FAQCategoryUpdate,
    FAQCategoryResponse,
    FAQCategoryListResponse,
    FAQCategoryWithFAQsResponse,
    FAQCategoryBulkAction,
    FAQCategoryReorderRequest
)
from app.schemas.base import StandardResponse
from app.services.faq_category_service import FAQCategoryService

router = APIRouter()
service = FAQCategoryService()


@router.get("", response_model=StandardResponse)
async def list_categories(
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    include_stats: bool = Query(True, description="Include FAQ counts and statistics")
):
    """
    Get all FAQ categories (public endpoint)
    For public use, only returns active categories by default
    """
    # For public API, default to active categories only
    if is_active is None:
        is_active = True
    
    result = await service.list_categories(is_active=is_active, include_stats=include_stats)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ categories retrieved successfully"
    )


@router.get("/admin", response_model=StandardResponse)
async def list_categories_admin(
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    include_stats: bool = Query(True, description="Include FAQ counts and statistics"),
    current_user: User = Depends(get_admin_user)
):
    """
    Get all FAQ categories (admin endpoint)
    Returns both active and inactive categories
    """
    result = await service.list_categories(is_active=is_active, include_stats=include_stats)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ categories retrieved successfully"
    )


@router.get("/{category_id}", response_model=StandardResponse)
async def get_category(category_id: str):
    """
    Get single FAQ category by ID (public endpoint)
    """
    result = await service.get_category(category_id)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ category retrieved successfully"
    )


@router.get("/{category_id}/faqs", response_model=StandardResponse)
async def get_category_with_faqs(category_id: str):
    """
    Get FAQ category with its FAQs (public endpoint)
    """
    result = await service.get_category_with_faqs(category_id)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ category with FAQs retrieved successfully"
    )


@router.post("", response_model=StandardResponse)
async def create_category(
    category_data: FAQCategoryCreate,
    current_user: User = Depends(get_admin_user)
):
    """
    Create new FAQ category (admin only)
    """
    result = await service.create_category(category_data)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ category created successfully"
    )


@router.put("/{category_id}", response_model=StandardResponse)
async def update_category(
    category_id: str,
    category_data: FAQCategoryUpdate,
    current_user: User = Depends(get_admin_user)
):
    """
    Update FAQ category (admin only)
    """
    result = await service.update_category(category_id, category_data)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ category updated successfully"
    )


@router.delete("/{category_id}", response_model=StandardResponse)
async def delete_category(
    category_id: str,
    current_user: User = Depends(get_admin_user)
):
    """
    Delete FAQ category (admin only)
    Cannot delete if category has FAQs
    """
    result = await service.delete_category(category_id)
    
    return StandardResponse(
        success=True,
        data=result,
        message=result["message"]
    )


@router.post("/reorder", response_model=StandardResponse)
async def reorder_categories(
    reorder_data: FAQCategoryReorderRequest,
    current_user: User = Depends(get_admin_user)
):
    """
    Bulk update category order (admin only)
    Expects: {"category_orders": [{"id": "category_id", "order": 0}, ...]}
    """
    result = await service.reorder_categories(reorder_data.category_orders)
    
    return StandardResponse(
        success=True,
        data=result,
        message=result["message"]
    )


@router.post("/bulk-action", response_model=StandardResponse)
async def bulk_action_categories(
    bulk_data: FAQCategoryBulkAction,
    current_user: User = Depends(get_admin_user)
):
    """
    Perform bulk actions on FAQ categories (Admin only)
    Actions: activate, deactivate, delete
    """
    result = await service.bulk_action(bulk_data.category_ids, bulk_data.action)
    
    return StandardResponse(
        success=True,
        data=result,
        message=result.get("message", "Bulk action completed successfully")
    )

File: frontend/lib/api/faq-categories.ts
/**
 * FAQ Categories API client
 * Following CLAUDE.md patterns - Dumb frontend, just call APIs
 */
import { apiClient } from './api-client';

export interface FAQCategoryData {
  id: string;
  name: string;
  slug: string;
  description?: string;
  platform_context?: string;  // Context for AI FAQ generation
  order: number;
  is_active: boolean;
  faq_count?: number;
  total_views?: number;
  has_faqs?: boolean;
  created_at: string;
  updated_at: string;
}

export interface FAQCategoryCreateData {
  name: string;
  slug: string;
  description?: string;
  order?: number;
  is_active?: boolean;
}

export interface FAQCategoryUpdateData {
  name?: string;
  slug?: string;
  description?: string;
  order?: number;
  is_active?: boolean;
}

export interface FAQCategoriesListResponse {
  categories: FAQCategoryData[];
  total: number;
  active_count: number;
  inactive_count: number;
}

export interface FAQCategoryWithFAQsResponse extends FAQCategoryData {
  faqs: any[]; // FAQ data
}

/**
 * FAQ Categories API client
 * Smart backend handles all business logic, frontend just calls endpoints
 */
export const faqCategoriesApi = {
  /**
   * List FAQ categories (public endpoint)
   * For public use, returns only active categories by default
   */
  list: (params?: { 
    is_active?: boolean; 
    include_stats?: boolean 
  }) => {
    const queryParams = new URLSearchParams();
    if (params?.is_active !== undefined) queryParams.append('is_active', params.is_active.toString());
    if (params?.include_stats !== undefined) queryParams.append('include_stats', params.include_stats.toString());
    
    return apiClient.get<StandardResponse<FAQCategoriesListResponse>>(
      `/faq-categories${queryParams.toString() ? `?${queryParams.toString()}` : ''}`
    );
  },

  /**
   * List FAQ categories (admin endpoint)
   * Returns both active and inactive categories
   */
  listAdmin: (params?: { 
    is_active?: boolean; 
    include_stats?: boolean 
  }) => {
    const queryParams = new URLSearchParams();
    if (params?.is_active !== undefined) queryParams.append('is_active', params.is_active.toString());
    if (params?.include_stats !== undefined) queryParams.append('include_stats', params.include_stats.toString());
    
    return apiClient.get<StandardResponse<FAQCategoriesListResponse>>(
      `/faq-categories/admin${queryParams.toString() ? `?${queryParams.toString()}` : ''}`
    );
  },

  /**
   * Get single FAQ category by ID
   */
  get: (id: string) => {
    return apiClient.get<StandardResponse<FAQCategoryData>>(
      `/faq-categories/${id}`
    );
  },

  /**
   * Get FAQ category with its FAQs
   */
  getWithFAQs: (id: string) => {
    return apiClient.get<StandardResponse<FAQCategoryWithFAQsResponse>>(
      `/faq-categories/${id}/faqs`
    );
  },

  /**
   * Create new FAQ category (admin only)
   */
  create: (data: FAQCategoryCreateData) => {
    return apiClient.post<StandardResponse<FAQCategoryData>>(
      '/faq-categories',
      data,
      { requireAuth: true }
    );
  },

  /**
   * Update FAQ category (admin only)
   */
  update: (id: string, data: FAQCategoryUpdateData) => {
    return apiClient.put<StandardResponse<FAQCategoryData>>(
      `/faq-categories/${id}`,
      data,
      { requireAuth: true }
    );
  },

  /**
   * Delete FAQ category (admin only)
   * Cannot delete if category has FAQs
   */
  delete: (id: string) => {
    return apiClient.delete<StandardResponse<{ success: boolean; message: string }>>(
      `/faq-categories/${id}`,
      { requireAuth: true }
    );
  },

  /**
   * Bulk update category order (admin only)
   */
  reorder: (orders: Array<{ id: string; order: number }>) => {
    return apiClient.post<StandardResponse<{ success: boolean; message: string }>>(
      '/faq-categories/reorder',
      { category_orders: orders },
      { requireAuth: true }
    );
  },

  /**
   * Perform bulk actions on FAQ categories (admin only)
   * Actions: activate, deactivate, delete
   */
  bulkAction: (categoryIds: string[], action: 'activate' | 'deactivate' | 'delete') => {
    return apiClient.post<StandardResponse<{
      success: boolean;
      message: string;
      affected: number;
      errors?: string[];
    }>>(
      '/faq-categories/bulk-action',
      { category_ids: categoryIds, action },
      { requireAuth: true }
    );
  }
};

// Type for standard API response
interface StandardResponse<T> {
  success: boolean;
  data: T;
  message: string;
  errors?: string[];
}